Add benchmarking performance test and run script

rjwills28 opened the pull request:
I have added the Python performance script (i.e. websocket client) that creates _N_ subscriptions to Coniql and measures memory usage, CPU and the number of missed updates. I have also added a bash script that takes care of creating the EPICS db, starts the IOC, starts Coniql and then runs _M_ instances of the Python ws client (to simulate multiple clients connecting to a single Coniql instance).

Results from the Python client script get appended to a file so that you can compare different runs.

The number of PVs to subscribe to, number of samples to collect and number of websocket clients are all configurable.

Do we want this version of the performance tests merged in as part of Coniql? They might be useful for individual developers to run locally when making changes. Otherwise the Python client can be used for the automated performance tests and we just use the bash script as a template.

AlexanderWells-diamond commented on benchmark/coniql_performance_test.py:
```diff
@@ -0,0 +1,341 @@
+                await websocket.send(request_message_graphql_ws_protocol)
+
+            msg_count = 0
+            async for response in websocket:
```
It'd be good to put a progress message somewhere in here - it's currently impossible to know how far through the test we are. Will need to do something funny with the output though, as this is in a thread the stdout will go elsewhere...

AlexanderWells-diamond commented on benchmark/run_performance_test.sh:
```diff
@@ -0,0 +1,151 @@
+
+# Some parameters
+SUB_DIR="benchmark"
+N_SAMPLES=1000
+# 1 = old, 2 = new
+PROTOCOL=2
```
Is it worth turning these into more parameters for the script? Its nicer to have everything you may want to change as a command line parameter rather than needing to edit the script itself.

rjwills28 replied:
I've improved the parsing of command line arguments in the script and also added these as parameters for the script.

AlexanderWells-diamond reviewed (commented):
Looks good and will definitely be helpful for locally checking for significant performance differences when changing Coniql. We should also be able to re-use large parts of this for a container-based performance test setup.

How long do these tests take for you to run? From my understanding the default is to create a bunch of records that add 1 every .1 seconds, and then the client is waiting for 1000 updates, which should take 100 seconds? I'm seeing test runs take 5+ minutes so am confused where the time is being used.

There seems to be a crash when you try and increase the number of PVs in use above 1000. 
I'm not sure what the start of the issue is but the last thing in the console is:
```
Exception in thread Thread-175:
Traceback (most recent call last):
-> Starting subscription: TEST:REC1806
  File "/usr/lib64/python3.8/threading.py", line 932, in _bootstrap_inner
  File "/usr/lib64/python3.8/threading.py", line 870, in run
  File "benchmark/coniql_performance_test.py", line 219, in coniql_subscription
  File "/usr/lib64/python3.8/asyncio/runners.py", line 44, in run
  File "/usr/lib64/python3.8/asyncio/base_events.py", line 616, in run_until_complete
  File "benchmark/coniql_performance_test.py", line 69, in subscribe
  File "/home/eyh46967/dev/coniql/venv_test1/lib64/python3.8/site-packages/websockets/imports.py", line 77, in __getattr__
  File "/home/eyh46967/dev/coniql/venv_test1/lib64/python3.8/site-packages/websockets/imports.py", line 28, in import_name
AttributeError: module 'websockets.legacy.client' has no attribute 'connect'
libgcc_s.so.1 must be installed for pthread_cancel to work
bash: line 1: 561763 Aborted                 (core dumped) python benchmark/coniql_performance_test.py -n 2000 -s 1000 -p 2 -f benchmark/performance_test_results_NClients_1.txt
```

I've got a few feedback points. This is an unsorted list and some of these points may not be worth acting upon.


- Who is the intended reader for the results file? It's fine for human readability, I'm just wondering if we want a slightly more computer-friendly format e.g. csv, json.
- Can there be a "total runtime" added to the end of the script? 
- Is there a way to estimate the remaining time? I notice that increasing the number of clients seems to exponentially increase the runtime which makes it hard to know if it's still running correctly or whether something might have broken
- I note that the `print` statements in the Python file don't seem to appear anywhere - it looks like that printing is never passed back to the stdout of the original shell? This is another thing that would make debugging issues somewhat tricky, especially as I'm assuming stderr is also not being passed along.
  - A bit more investigation revealed that this is actually an issue with using VSCode's inbuilt terminal - it seems that it doesn't actually display the new tabs, they're created in the background somewhere. They still run, we just don't see the terminals and so we don't see the output. Do you know if there's a way to detect this from inside the shell?
- It's tricky to spot errors in the spawned tabs, as they close at the end of the run. This is easy to recreate if you try running the script outside of the Coniql venv - there will be errors in the terminals but if you wait too long they close and lose the information
- Can you make it so that, even if the Python script exits unexpectedly, it will still try to write out whatever results it has collected? Its an irritation that we can just lose all data in some cases

AlexanderWells-diamond commented:
Another point to consider: We've still got the old Python and Javascript benchmarking scripts (along with some docs on how to run them) still present. Do we want to get rid of them? Should we add this script to the docs?

rjwills28 commented:
Thanks for the detailed review. I'll look at addressing the comments and additional features to the scripts.
I must admit I had not tried running with 1000 PVs. I see my CPU hit 100% when running with 200 PVs so that was as far as I went. I assumed that the performance tests will be run against a single instance of Coniql, which means it will have limits so we might have to think about what a reasonable number is to stress the system and still get meaningful results.

Quick comment on the runtime of the performance test. It runs until it has captures N samples (e.g. 1000 in your example). When you increase the number of PVs, you increase the CPU usage. When you hit 100% CPU then Coniql suffers and starts to miss PV updates. This means it takes longer for us to collect 1000 samples as more updates have been missed, if that makes sense.

In terms of the 'old' benchmarking scripts, I haven't looked at them myself but if we don't think they will be used once we have these tests running then I think we should remove them. Adding instructions on how to run these new scripts should then be added to the documentation.

AlexanderWells-diamond commented on benchmark/run_performance_test.sh:
```diff
@@ -0,0 +1,151 @@
+echo "-> Creating EPICS db with $N_PVS PVs"
+for ((i=0;i<$N_PVS;i++))
+do
+    if [ $i -lt 10 ]; then 
+        str_name="0$i"
+    else
+        str_name="$i"
+    fi
```
Further testing has revealed that this line is causing problems - if I set it to `"%03d"` it works, but using `05` and it doesn't. I'll investigate why this is.

AlexanderWells-diamond commented on benchmark/coniql_performance_test.py:
```diff
@@ -237,46 +234,38 @@ def main():
+    cpu_monitor_thread = threading.Thread(target=cpu_monitor, args=(signal,))
+    cpu_monitor_thread.start()
+
+    # Create subsciption tasks for n_pvs
```
```suggestion
    # Create subscription tasks for n_pvs
```

rjwills28 replied:
Fixed.

rjwills28 commented on benchmark/coniql_performance_test.py:
```diff
@@ -0,0 +1,341 @@
+import threading
+import time
+
+import psutil
+import websockets
```
I have included these dependencies in the `pyproject.toml` and can then use the same venv to run the performance tests instead of having to create a new one and tidy up.

rjwills28 commented on benchmark/coniql_performance_test.py:
```diff
@@ -0,0 +1,341 @@
+                        else:
+                            if msg_count > n_messages:
+                                break
+                                print("break")
```
Good catch, thanks.

rjwills28 commented on benchmark/run_performance_test.sh:
```diff
@@ -0,0 +1,151 @@
+pids=$(pgrep softIoc)
+kill -INT $pids
+pids=$(pgrep coniql)
+kill -INT $pids
```
I've added a check to see whether the process is still running before attempting to kill it, which should stop such errors.

rjwills28 commented on benchmark/coniql_performance_test.py:
```diff
@@ -0,0 +1,341 @@
+                client,
+                pv_name,
+                n_samples,
+            ),
```
I've refactored to use `asyncio` instead of threads.

rjwills28 commented:
Addressing some of the general comments:

> Who is the intended reader for the results file? It's fine for human readability, I'm just wondering if we want a slightly more computer-friendly format e.g. csv, json.

I'm not sure about this one. I assume that we would want to look at the results and compare different runs ourselves. My intention was to write the results to file so that they can be extracted and presented in some way but I'm not sure how you would plan to do that? Perhaps in that case it would be better if they were just raw numbers that were then interpreted by some other method?

> Can there be a "total runtime" added to the end of the script?

I have added the total runtime to the end of the script.

> Is there a way to estimate the remaining time? I notice that increasing the number of clients seems to exponentially increase the runtime which makes it hard to know if it's still running correctly or whether something might have broken

I have added some progress monitoring to the Python script and also estimate the remaining time based on how long it has already taken to collect x samples.

> I note that the print statements in the Python file don't seem to appear anywhere - it looks like that printing is never passed back to the stdout of the original shell? This is another thing that would make debugging issues somewhat tricky, especially as I'm assuming stderr is also not being passed along.

I have configured it so that the stdout & stderr of EPICS, Coniql and the performance test are written to log files so that we can always grab the output from the latest run. These will be saved under `/benchmark/logs`.

> It's tricky to spot errors in the spawned tabs, as they close at the end of the run. This is easy to recreate if you try running the script outside of the Coniql venv - there will be errors in the terminals but if you wait too long they close and lose the information
    
See above comment.

> Can you make it so that, even if the Python script exits unexpectedly, it will still try to write out whatever results it has collected? Its an irritation that we can just lose all data in some cases

Do you mean even if it doesn't complete. I have added a `try-except` block to run the subscriptions so even if this fails we should still reach the results write out code.

AlexanderWells-diamond reviewed (changes requested):
Looking better and better. Very nearly there. A couple of minor things, and an issue with the `progress.txt` file, to resolve.

As an aside, I'm wondering if it wouldn't be better/easier at this point to have done everything inside a Python script. Using the `subprocess` module would I think significantly ease some of the headaches of reading output from files. Although a rewrite at this stage probably isn't worth it. I'll see what I want/need to do to get this running well in Kubernetes before suggesting anything more significant.

AlexanderWells-diamond commented on benchmark/coniql_performance_test.py:
```diff
@@ -0,0 +1,366 @@
+        return self.values
+
+
+def cpu_monitor(signal):
```
A minor niggle: Every instance of this client program will spawn this same CPU monitor. All but the first are then giving redundant information. The right fix is probably to turn this into a standalone program, but at this point that may not be worth the effort.

AlexanderWells-diamond replied:
Attaching creating this to the logging flag should work.

rjwills28 replied:
I have added a separate argument for the Python script named `--no-cpu-monitor`. If this is provided then the CPU monitor will not be started. I thought that the default behaviour should be that the CPU monitor runs, even if a log file for the progress is not provided and so I added the extra argument.
Within the bash script, the first client specifies a log file so will monitor subscription progress and will monitor the CPU, while subsequent clients will not run a CPU monitor.

rjwills28 commented on benchmark/run_performance_test.sh:
```diff
@@ -0,0 +1,229 @@
+        fi
+    else
+        running=true
+        tail=$(tail --lines=1 /tmp/progress.txt)
+        if [ "${tail}" != "${progress}" ]; then
+            progress="${tail}"
+            echo "   "$progress
+        fi
```
Yes this doesn't work with multiple clients but in the bash script I have fixed this by calling the first instance of the Python clients with the `-l` argument signalling the first client to log progress. I.e. https://github.com/DiamondLightSource/coniql/blob/1276ac5ce73815488e4490008769ed7820160dd2/benchmark/run_performance_test.sh#L182

Subsequent clients are not started with this and so do not open or log to the file, e.g. https://github.com/DiamondLightSource/coniql/blob/1276ac5ce73815488e4490008769ed7820160dd2/benchmark/coniql_performance_test.py#L66

rjwills28 commented:
> As an aside, I'm wondering if it wouldn't be better/easier at this point to have done everything inside a Python script. 

At some point the idea of using Jenkins to run these tests was thrown around and so I went down the bash route. Just let me know if you think we do need to refactor.

rjwills28 commented on benchmark/coniql_performance_test.py:
```diff
@@ -45,15 +46,32 @@
+            self.progress_file = open("/tmp/progress.txt", "w")
 
     async def subscribe(self, idid, query, handle, n_messages):
+        monitor_progress = False
```
I think it is needed. We have 2 cases that we need to protect against to make sure that there is only a single instance that logs progress:
1. In a single client there will be multiple `subscribe` calls - one for each subscription to a PV. We only need one of the subscriptions to log progress so I have picked the first one. The first `subscribe` will have `self.first_subscribe` set to `True` and so will enter the case statement. We then need to immediately switch `self.first_subscribe` to `False` so that subsequent subscribes don't trigger the logging. That means that we need the `monitor_progress` variable to save the fact that we should log.
2. Our bash script has the option to create multiple clients but we only want a single client to be able to log progress. Therefore I introduced the `self.log` parameter, which is passed in as a variable to the Python script and sets it to `True` for the first client (indicating it should log progress) and `False` for all other instances.

AlexanderWells-diamond replied:
Ah right, I had failed to see it was the same `GraphQLClient` for every subscription. Yes, with the current setup that does then make sense.

rjwills28 commented on benchmark/coniql_performance_test.py:
```diff
@@ -0,0 +1,366 @@
+    try:
+        await asyncio.gather(*task_list)
+        # Signal to CPU monitoring thread to stop recording CPU metrics
+        signal.signal_stop()
```
Ah so have the creation of the tasks in the `try` block and the `await ayncio.gather` in the `finally` should something go wrong in the creation?

AlexanderWells-diamond replied:
No, sorry, I just meant the `signal.signal_stop()` should be in the `finally` block.

rjwills28 replied:
Thanks for clarifying - that makes sense. I have moved this to the `finally` block.

rjwills28 commented on benchmark/coniql_performance_test.py:
```diff
@@ -0,0 +1,366 @@
+        )
+
+    # Collect results
+    time.sleep(1)
```
Once we have gathered all subscriptions we signal the CPU monitoring thread to stop recording CPU metrics. The CPU monitor is running with a `while` loop that will break when it receives this signal, but the while loop has a `time.sleep(0.1)` between each iteration so potentially there is a 0.1 sec delay in signalling a stop and the loop actually stopping. The `sleep(1)` is there to just ensure the loop has had time to exit.

AlexanderWells-diamond commented on benchmark/coniql_performance_test.py:
```diff
@@ -45,15 +46,32 @@
+
+        if self.log:
+            print("Logging subscription progress")
+            self.progress_file = open("/tmp/progress.txt", "w")
```
Yes, perhaps turn the `--log` argument into a `--log-file` and pass it the full file path of a newly created directory and file in the `/tmp` directory, similar to how `pytest` does it with the creation of a new folder including a timestamp for each run.

AlexanderWells-diamond reviewed (approved):
Looking good!

rjwills28 commented:
Thanks @AlexanderWells-diamond for all the help and suggesting on this PR.
